django/db/models/lookups.py

```diff
@@ -569,10 +582,22 @@ def as_sql(self, compiler, connection):
                 "The QuerySet value for an isnull lookup must be True or False."
             )
         sql, params = compiler.compile(self.lhs)
+        placeholder = "%s"
+        # TODO: We cannot use as_postgresql because RelatedIsNull etc
+        # doesn't pick this up ://////////////////////////////////////
+        # Subclassed lookups are a PITA
+        if connection.vendor == "postgresql":
+            # TODO: What is the proper way to figure out the cast type?
+            if not isinstance(self.lhs, Col) and hasattr(
+                self.lhs.field, "cast_db_type"
+            ):
+                cast_db_type = self.lhs.field.cast_db_type(connection)
+                if cast_db_type:
+                    placeholder = "(%s)::" + cast_db_type
         if self.rhs:
-            return "%s IS NULL" % sql, params
+            return (placeholder + " IS NULL") % sql, params
         else:
-            return "%s IS NOT NULL" % sql, params
+            return (placeholder + " IS NOT NULL") % sql, params
```

Comment:
I suspect there is a way to entirely avoid that by making sure the proper `output_field` is assigned and adapted for the backend.

Reply:
@charettes I looked into this and it doesn't seem to be that easy. The `output_field` is assigned properly enough which leads me down into: 

https://github.com/django/django/blob/2848e5d0ce5cf3c31fe87525536093b21d570f69/django/db/models/expressions.py#L1011-L1027

`get_placeholder` would help here but is not defined for standard types :/ Would we need a connection feature `requires_cast_for_value` or do you have something else in mind? Also do we always need to cast or only when we are on the lhs?

Reply:
Fix by my latest commits :)